beater/beater.go

```diff
@@ -310,63 +332,75 @@ type sharedServerRunnerParams struct {
 }
 
 func newServerRunner(ctx context.Context, args serverRunnerParams) (*serverRunner, error) {
-	cfg, err := config.NewConfig(args.RawConfig, elasticsearchOutputConfig(args.Beat))
+	runServerContext, cancel := context.WithCancel(ctx)
+	config, err := config.NewConfig(args.RawConfig, elasticsearchOutputConfig(args.Beat))
 	if err != nil {
 		return nil, err
 	}
-
-	runServerContext, cancel := context.WithCancel(ctx)
-	return &serverRunner{
+	s := &serverRunner{
 		backgroundContext:      ctx,
 		runServerContext:       runServerContext,
 		cancelRunServerContext: cancel,
-
-		config:        cfg,
-		rawConfig:     args.RawConfig,
-		fleetConfig:   args.FleetConfig,
-		acker:         args.Acker,
-		pipeline:      args.Pipeline,
-		namespace:     args.Namespace,
-		beat:          args.Beat,
-		logger:        args.Logger,
-		tracer:        args.Tracer,
-		tracerServer:  args.TracerServer,
-		wrapRunServer: args.WrapRunServer,
-	}, nil
-}
-
-func (s *serverRunner) String() string {
-	return "APMServer"
+		config:                 config,
+		rawConfig:              args.RawConfig,
+		fleetConfig:            args.FleetConfig,
+		acker:                  args.Acker,
+		pipeline:               args.Pipeline,
+		namespace:              args.Namespace,
+		beat:                   args.Beat,
+		logger:                 args.Logger,
+		tracer:                 args.Tracer,
+		tracerServer:           args.TracerServer,
+		wrapRunServer:          args.WrapRunServer,
+		restartc:               make(chan struct{}),
+		// This is just for testing right now.
+		params: args,
+	}
```

Comment:
there's a lot of duplication here between setting the initial state and configuring it. ideally this could be minimized. i believe i added it so that configure wouldn't have to do a bunch of nil checks on the incoming config, as well as the struct members on the serverRunner. Could maybe have newServerRunner just return a default serverRunner with default values for all its struct members, which would eliminate needing to pass in arguments.